phoenix-core-server/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java

```diff
@@ -1652,10 +1717,37 @@ private List<Mutation> generateOnDupMutations(BatchMutateContext context, Put at
               delete.add(cell);
           }
       }
+
+      if (put != null) {
+          // if put is empty and delete is null, remove the empty put from mutations,
+          // otherwise we should also add empty cell timestamp for update
+          if (put.isEmpty() && delete == null) {
+              mutations.remove(put);
+          } else {
+              PTable table = operations.get(0).getFirst();
+              addEmptyKVCells(put, delete, tuple, table);
+          }
+      }
+
       return mutations;
   }
 
-
+    private void addEmptyKVCells(Put put, Delete delete, MultiKeyValueTuple tuple, PTable table) throws IOException {
+        Set<byte[]> familySet = new HashSet<>();
+        if (!put.isEmpty()) {
+            familySet.addAll(put.getFamilyCellMap().keySet());
+        }
+        if (delete != null) {
+            familySet.addAll(delete.getFamilyCellMap().keySet());
+        }
+        for (byte[] familyBytes : familySet) {
```

Comment:
Shouldn't we use this to get empty column family `byte[] emptyCF = SchemaUtil.getEmptyColumnFamily(table)`

Reply:
Done